app/vmalert/rule/alerting.go

```diff
@@ -313,11 +313,8 @@ func (ar *AlertingRule) execRange(ctx context.Context, start, end time.Time) ([]
 	}
 	var result []prompbmarshal.TimeSeries
 	holdAlertState := make(map[uint64]*notifier.Alert)
-	qFn := func(_ string) ([]datasource.Metric, error) {
-		return nil, fmt.Errorf("`query` template isn't supported in replay mode")
-	}
 	for _, s := range res.Data {
-		ls, as, err := ar.expandTemplates(s, qFn, time.Time{})
+		ls, as, err := ar.expandTemplates(s, nil, time.Time{})
```

Comment:
This commit contains hidden change - it removes `query` function from `query_range` aka `vmalert-replay` mode. 
 
 WDYT of replacing it with `no-op` function? Because after this change, it may require to rewrite all alerting rules, before passing to the `replay-mode`. Which makes it hard to use.

Reply:
The behavior doesn't change, query func by default return a similar error and is no-op.
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/3513eb04bdc78ee9983156dd970f030884424b8b/app/vmalert/templates/template.go#L149-L151
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/3513eb04bdc78ee9983156dd970f030884424b8b/app/vmalert/notifier/alert.go#L154-L158

Reply:
I'll create an issue for it. Looks like it's not possible to use replay mode with `query` function.